### PR TITLE
[FIX] visibility errors on gcc - alphabet

### PR DESCRIPTION
**Pull request by marehr**
Superseeds #1228

fixes partially #1317

### CODE REVIEW

File: include/seqan3/alphabet/composite/alphabet_tuple_base.hpp [hunk 1]
```diff
@@ -674,7 +674,7 @@ template <typename indirect_component_type, typename derived_type, typename ...i
              !detail::weakly_ordered_by_members_with<indirect_component_type, derived_type>
 //!\endcond
 constexpr bool operator>=(indirect_component_type const lhs,
-                          alphabet_tuple_base<derived_type, indirect_component_types...> const rhs) noexcept
+                          alphabet_tuple_base<derived_type, indirect_component_types...> const & rhs) noexcept
 {
```
Comment (h-2) [inline include/seqan3/alphabet/composite/alphabet_tuple_base.hpp]: These won't be necessary if #1297 gets merged...

File: include/seqan3/alphabet/concept.hpp [hunk 2]
```diff
@@ -618,7 +618,7 @@ void alphabet_size(args_t ...) = delete;
 template <typename alph_t>
 struct alphabet_size_fn
 {
-private:
+public:
```
Comment (h-2) [inline include/seqan3/alphabet/concept.hpp]: These are weird, but oh well.

File: include/seqan3/alphabet/detail/alphabet_proxy.hpp [hunk 3]
```diff
@@ -263,38 +263,41 @@ class alphabet_proxy : public
      *        e.g. seqan3::alphabet_variant.
      * \{
      */
+private:
+    //!\brief work around a gcc bug that disables short-circuiting of operator&& in an enable_if_t of a friend function
+    template <typename t>
+    static constexpr bool is_alphabet_comparable_with = !std::is_same_v<derived_type, t> &&
+                                                        detail::weakly_equality_comparable_with<alphabet_type, t>;
+
```
Comment (h-2) [inline include/seqan3/alphabet/detail/alphabet_proxy.hpp]: This makes a difference? Interesting.

Comment (marehr) [reply include/seqan3/alphabet/detail/alphabet_proxy.hpp]: Yes it does. Maybe this fixes your problems where short circuiting didn't work.

Comment (h-2) [reply include/seqan3/alphabet/detail/alphabet_proxy.hpp]: I moved all the checks out into external variables anyway so I implicitly did the same thing, I think.

File: include/seqan3/alphabet/concept.hpp (hunk 2 above)
Comment (rrahn) [reply include/seqan3/alphabet/concept.hpp]: But do we know why this is necessary? I mean I don't see right now why the private CPO members are not accessible from within the class itself?

File: test/unit/alphabet/nucleotide/sam_dna16_test.cpp [hunk 4]
```diff
@@ -25,6 +25,11 @@ INSTANTIATE_TYPED_TEST_CASE_P(sam_dna16, semi_alphabet_test, sam_dna16);
 INSTANTIATE_TYPED_TEST_CASE_P(sam_dna16, alphabet_constexpr, sam_dna16);
 INSTANTIATE_TYPED_TEST_CASE_P(sam_dna16, semi_alphabet_constexpr, sam_dna16);
 
+struct sam_dna16_accessor : public sam_dna16
+{
+    using sam_dna16::rank_to_char;
```
Comment (rrahn) [inline test/unit/alphabet/nucleotide/sam_dna16_test.cpp]: so this is protected in sam_dna_16, but we have a test that tests this conversion of the sam?
Should that test not use the public interface to check the correctness of the conversion? I mean I am asking myself why it is protected and not private?

File: include/seqan3/alphabet/concept.hpp (hunk 2 above)
Comment (marehr) [reply include/seqan3/alphabet/concept.hpp]: clang allows this, gcc not. Not sure what the standard says, I would expect it to work, but it doesn't.

Comment (marehr) [reply include/seqan3/alphabet/concept.hpp]: I asked on the concepts [slack](https://cpplang.slack.com/archives/C6K47F8TT/p15674338610107000), but none answered.

Comment (h-2) [reply include/seqan3/alphabet/concept.hpp]: It works with GCC10 with c++20 concepts and concepts-ts. I would consider it a bug of GCC<=9, but one that likely won't be fixed.
Nothing to be done about it.

Comment (h-2) [reply include/seqan3/alphabet/concept.hpp]: (You can WORKAROUND declare it for GCC<=9 but that arguably makes the code worse that just making it public, I think)

Comment (marehr) [reply include/seqan3/alphabet/concept.hpp]: okay, good to know

Comment (rrahn) [reply include/seqan3/alphabet/concept.hpp]: 💥

File: test/unit/alphabet/detail/alphabet_proxy_test.cpp [hunk 5]
```diff
@@ -120,6 +128,10 @@ class alphabet_proxy_example2 : public alphabet_proxy<alphabet_proxy_example2, m
     constexpr alphabet_proxy_example2 & operator=(alphabet_proxy_example2 const &) = default;
     constexpr alphabet_proxy_example2 & operator=(alphabet_proxy_example2 &&) = default;
     ~alphabet_proxy_example2() = default;
+
+    constexpr alphabet_proxy_example2(alphabet_type const a) noexcept
+        : base_t{a}
```
Comment (rrahn) [inline test/unit/alphabet/detail/alphabet_proxy_test.cpp]: Why is this not on one line? I see no need to wrap this? Also when wrapping I think we agreed on putting the operator on the previous line and then continue the remaining expression on the next except for the pipe-notation for the ranges.

File: test/unit/alphabet/detail/alphabet_proxy_test.cpp [hunk 6]
```diff
@@ -31,6 +32,12 @@ class alphabet_proxy_example : public alphabet_proxy<alphabet_proxy_example, dna
     constexpr alphabet_proxy_example & operator=(alphabet_proxy_example const &) = default;
     constexpr alphabet_proxy_example & operator=(alphabet_proxy_example &&) = default;
     ~alphabet_proxy_example() = default;
+
+    constexpr alphabet_proxy_example(alphabet_type const a) noexcept
+        : base_t{a}
```
Comment (rrahn) [inline test/unit/alphabet/detail/alphabet_proxy_test.cpp]: on the previous line?

File: test/unit/alphabet/detail/alphabet_proxy_test.cpp (hunk 5 above)
Comment (marehr) [reply test/unit/alphabet/detail/alphabet_proxy_test.cpp]: can you give an example?

**Pull request merged**